src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
I kept it with a fallback similar to some other code I've seen when people created the grains with some factory for their unit tests. I could revert it though, as you are right that I don't know the implications. Still, it doesn't seem unsafe, does it? I'm not ever disposing this. only the scope if set.
In the future, I also envision a way to configure it so that users are not forced to use scoped containers if they don't require it. I could delay this kind of change until then if you prefer.